libpod/container_api.go

```diff
@@ -847,31 +843,57 @@ func (c *Container) ShouldRestart(ctx context.Context) bool {
 	return c.shouldRestart()
 }
 
-// ResolvePath resolves the specified path on the root for the container.  The
-// root must either be the mounted image of the container or the already
-// mounted container storage.
-//
-// It returns the resolved root and the resolved path.  Note that the path may
-// resolve to the container's mount point or to a volume or bind mount.
-func (c *Container) ResolvePath(ctx context.Context, root string, path string) (string, string, error) {
-	logrus.Debugf("Resolving path %q (root %q) on container %s", path, root, c.ID())
+// CopyFromArchive copies the contents of the specified tarStream to path.
+func (c *Container) CopyFromArchive(ctx context.Context, path string, tarStream io.Reader) (func() error, error) {
+	if !c.batched {
+		c.lock.Lock()
+		defer c.lock.Unlock()
 
-	// Minimal sanity checks.
-	if len(root)*len(path) == 0 {
-		return "", "", errors.Wrapf(define.ErrInternal, "ResolvePath: root (%q) and path (%q) must be non empty", root, path)
+		if err := c.syncContainer(); err != nil {
+			return nil, err
+		}
 	}
-	if _, err := os.Stat(root); err != nil {
-		return "", "", errors.Wrapf(err, "cannot locate root to resolve path on container %s", c.ID())
+
+	return c.copyFromArchive(ctx, path, tarStream)
+}
+
+// CopyToArchive copies the contents of the specified path to the tarStream.
+func (c *Container) CopyToArchive(ctx context.Context, path string, tarStream io.Writer) (func() error, error) {
+	if !c.batched {
+		c.lock.Lock()
+		defer c.lock.Unlock()
+
+		if err := c.syncContainer(); err != nil {
+			return nil, err
+		}
 	}
 
+	return c.copyToArchive(ctx, path, tarStream)
+}
+
+// Stat the specified path and return file info.
+func (c *Container) Stat(ctx context.Context, path string) (*define.FileInfo, error) {
 	if !c.batched {
 		c.lock.Lock()
 		defer c.lock.Unlock()
```

Comment:
Should we consider making the container unlock after initial checks? Right now, `podman ps` while copying will block, which is not good - but at the same time, a container that stops while we are copying would be able to unmount the container from under us.

Reply:
I think it's okay for `Stat` since it's fast, but I will look into optimizing it for the two copy cases.